src/main/java/com/ably/kafka/connect/offset/OffsetRegistryService.java

```diff
@@ -0,0 +1,80 @@
+package com.ably.kafka.connect.offset;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.connect.sink.SinkRecord;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * This class is responsible for keeping track of the latest offset for each topic partition.
+ */
+public class OffsetRegistryService implements OffsetRegistry {
+
+    // Hashmap of TopicPartition to Offsets.
+    private final Map<TopicPartition, Long> topicPartitionToOffsetMap = new ConcurrentHashMap<>();
+
+    public OffsetRegistryService() {
+
+    }
+    @Override
+    public void update(TopicPartition partition, long offset) {
+
+    }
+
+    /**
+     * Returns the latest offset for a given topic partition.
+     * @param partition
+     * @return Offset
+     */
+    @Override
+    public long latestOffset(TopicPartition partition) {
+        return topicPartitionToOffsetMap.get(partition);
+    }
+
+    /**
+     * Updates the topicPartitionToOffsetMap with the latest offset for each topic partition.
+     * @param records
+     */
+    @Override
+    public void updateTopicPartitionToOffsetMap(List<SinkRecord> records) {
+        for (SinkRecord record : records) {
+            TopicPartition topicPartition = new TopicPartition(record.topic(), record.kafkaPartition());
+
+            if (topicPartitionToOffsetMap.containsKey(topicPartition)) {
+                if (topicPartitionToOffsetMap.get(topicPartition) < record.kafkaOffset()) {
+                    topicPartitionToOffsetMap.put(topicPartition, record.kafkaOffset());
```

Comment:
I think these independent calls to `containsKey()`, `get()` and `put()` can result in a lost update race condition. It's possible that competing threads modifying the offsets map after different batch responses will cause us to record an offset smaller than the latest committed record, e.g.

```
// topicPartitionToOffsetMap = {"partition_1" => 10}

Thread 1: if (topicPartitionToOffsetMap.get("partition_1") < 11) // true
Thread 2: topicPartitionToOffsetMap.put("partition_1", 15);
Thread 1: topicPartitionToOffsetMap.put("partition_1", 12); // update (partition_1 => 15) is lost
```

You might want to look into using `ConcurrentHashMap.compute()` or `ConcurrentHashMap.computeIfAbsent()` to make these read-update operations fully atomic.